Improves I/O performance by reading multiple seconds (files) into memory, in parallel

dipietrantonio opened the pull request:
This is a first implementation of an extra buffer between vcsbeam's `vmReadNextSecond` and MWALIB's `mwalib_voltage_context_read_second`. At this point it is still very much experimental, but I have tested it with the Crab dataset and it works. But it is too small to see any benefit.

What has been implemented:

- An extra buffer to read multiple seconds in memory at once.
- Files, especially MWAX ones, are read in one go.
- Parallelisation through OpenMP of reading the next `seconds_to_read` seconds.

dipietrantonio committed:
First implementation of a drop-in replacement function to read seconds of data.

dipietrantonio committed:
Adds OpenMP support for parallel reads of multiple seconds.

dipietrantonio committed:
Increases number of seconds in buffer.

dipietrantonio committed:
Code cleanup.

dipietrantonio committed:
Refactoring of the memory buffer implementation.

- Mainly to better integrate it with the existing codebase.
- Adds command line option to enable feature.

dipietrantonio committed:
Removes unused stopwatch.

bwmeyers commented:
Hey @dipietrantonio for some reason this fell off my plate. Looks like there is a conflict in the actual application code, but I'll review the other parts and then you can fix it (and any comments/suggestions) together.

bwmeyers commented on src/metadata.c:
```diff
@@ -922,6 +940,86 @@ void vmPushChunk( vcsbeam_context *vm )
 
+
+/**
+    @Cristian: here is a simple test to see if we can read as much data as we can in one go.
```
Can you clarify and/or adjust this comment block - it sounds like a test case at the moment, but is the actual implementation?

bwmeyers commented on src/metadata.c:
```diff
@@ -922,6 +940,86 @@ void vmPushChunk( vcsbeam_context *vm )
+                    size_t buffer_len,
+                    const char *error_message,
+                    size_t error_message_length){
+    
```
Since you access `vm->seconds_buffer` over and over, is there an easy C-syntax way of essentially saying:
```C
vmsb = vm->seconds_buffer;
...
```
so that then subsequent accesses/updates are easier to read?

dipietrantonio replied:
The `vm->seconds_buffer` object is the structure implementing the caching mechanism / buffer.

One thing that could be done is to define a pointer to it. Something like, `SecondsBuffer *vmsb = &(vm->seconds_buffer)`. Then, statements like  `vm->seconds_buffer.current_timestep_idx` could be shortened to `vmsb->current_timestep_idx`.

The reason I kept `vm->seconds_buffer` is to make it explicit where the structure is, but I can appreciate there is a lot of repetition.

Do you want me to implement such modification?

bwmeyers replied:
I think as long as it's clear, and even if you just add the comment that you're making that pointer to reduce repetition it will be useful.

bwmeyers commented on src/metadata.c:
```diff
@@ -922,6 +940,86 @@ void vmPushChunk( vcsbeam_context *vm )
+        vm->seconds_buffer.capacity = vm->seconds_per_file * (desired_seconds_in_buffer / vm->seconds_per_file);
+        size_t total_bytes = vm->bytes_per_second * vm->seconds_buffer.capacity;
+       
+        sprintf(vm->log_message, "read_next_second_from_buffer: will allocate %.4f GiB for 'seconds_buffer', "
```
I always mess this up, but just checking that this is GiB not GB?

dipietrantonio replied:
Division by `1024^3` -> GiB
Division by `1000^3` -> GB

bwmeyers commented on src/metadata.c:
```diff
@@ -922,6 +940,86 @@ void vmPushChunk( vcsbeam_context *vm )
+        size_t end_timestep_idx = vm->seconds_buffer.current_timestep_idx + n_timesteps_to_read;
+        size_t read_size = ((size_t) vm->bytes_per_second) * vm->seconds_per_file;
+        logger_start_stopwatch( vm->log, "read-seconds", true);
+        // sprintf(vm->log_message, "read_next_second_from_buffer: will now read %lu seconds into the buffer.\n", seconds_to_read);
```
Can remove comment print line(s)?

dipietrantonio replied:
Well spotted, I was thinking of logging every time the buffer is refilled but then realised it might create unnecessary output..

bwmeyers commented on src/metadata.c:
```diff
@@ -922,6 +940,86 @@ void vmPushChunk( vcsbeam_context *vm )
+        logger_message( vm->log, vm->log_message );
+        // Read multiple (file) timesteps in parallel using OpenMP
+        #pragma omp parallel for schedule(static)
+        for(size_t timestep_idx = start_timestep_idx; timestep_idx < end_timestep_idx; timestep_idx += 1){
```
Some interesting formatting/style here... maybe a little bit of a clean up? We might want to Lint the entire code base once we've settled on feature implementations later...

dipietrantonio replied:
What are you referring to, exactly? The extra whitespaces in `logger_message( vm->log, vm->log_message );` ? I will remove those, but I think I copy pasted it from another part of the code..

bwmeyers replied:
I think, in the preview on GitHub, there was rather a lot of white space all over the place but perhaps that was actually just the web rendering.

bwmeyers commented on src/metadata.c:
```diff
@@ -922,6 +940,86 @@ void vmPushChunk( vcsbeam_context *vm )
+
+
 /**
  * Reads a second's worth of input data from the observation files.
```
I suppose this comment isn't quite true any more? It reads from the buffer?

bwmeyers replied:
Ah I see, only uses the buffer if asked to, otherwise DOES read directly from the file.

dipietrantonio replied:
Yes, exactly as you observed. I the function still reads one second of data. The buffer implementation is transparent to the caller, hence why I did not modify the description of the function.

bwmeyers reviewed (changes requested):
Looks okay to me. Just a few clarifications/small fixes. Plus the apparent merge conflict to resolve...

dipietrantonio commented:
Thank you for the review, Bradley. I will work on it sometime this week.

dipietrantonio committed:
Improves comments describing the buffer implementation.

dipietrantonio committed:
Merge branch 'main' into cdp-improve-reading

dipietrantonio commented:
> Hey @dipietrantonio for some reason this fell off my plate. Looks like there is a conflict in the actual application code, but I'll review the other parts and then you can fix it (and any comments/suggestions) together.

The conflict was due to the options string modified both by my branch, adding `Z:`, and the main branch, adding the `k` option.

dipietrantonio committed:
Make seconds_buffer a named struct.

dipietrantonio committed:
Improves readability of read_next_second_from_buffer.

..by substituting references to `vm->seconds_buffer` with a pointer to the same structure, `vmsb`.

bwmeyers commented on app/make_mwa_tied_array_beam.c:
I realise you didn't change it here, but there's a missing `\n` on this line we should add...

bwmeyers reviewed (approved):
Approved, thanks!

One tiny cosmetic thing (that wasn't even your change, I just realised it in the CLI options when looking at your change).

dipietrantonio committed:
Adds line end to help message and changes sprintf format string for size_t.
